packages/fetchai/connections/prometheus/connection.py

```diff
@@ -0,0 +1,341 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Prometheus connection and channel."""
+
+import asyncio
+import logging
+from typing import Any, Dict, Optional, Tuple, Union, cast
+
+import aioprometheus  # type: ignore
+
+from aea.common import Address
+from aea.configurations.base import PublicId
+from aea.connections.base import Connection, ConnectionStates
+from aea.exceptions import enforce
+from aea.mail.base import Envelope, EnvelopeContext, Message
+from aea.protocols.dialogue.base import Dialogue as BaseDialogue
+
+from packages.fetchai.protocols.prometheus.dialogues import PrometheusDialogue
+from packages.fetchai.protocols.prometheus.dialogues import (
+    PrometheusDialogues as BasePrometheusDialogues,
+)
+from packages.fetchai.protocols.prometheus.message import PrometheusMessage
+
+
+_default_logger = logging.getLogger("aea.packages.fetchai.connections.prometheus")
+
+PUBLIC_ID = PublicId.from_str("fetchai/prometheus:0.1.0")
+
+DEFAULT_HOST = "127.0.0.1"
+DEFAULT_PORT = 9090
+
+
+class PrometheusDialogues(BasePrometheusDialogues):
+    """The dialogues class keeps track of all prometheus dialogues."""
+
+    def __init__(self, **kwargs) -> None:
+        """
+        Initialize dialogues.
+
+        :return: None
+        """
+
+        def role_from_first_message(  # pylint: disable=unused-argument
+            message: Message, receiver_address: Address
+        ) -> BaseDialogue.Role:
+            """Infer the role of the agent from an incoming/outgoing first message
+
+            :param message: an incoming/outgoing first message
+            :param receiver_address: the address of the receiving agent
+            :return: The role of the agent
+            """
+            # The server connection maintains the dialogue on behalf of the agent
+            return PrometheusDialogue.Role.SERVER
+
+        BasePrometheusDialogues.__init__(
+            self,
+            self_address=str(PUBLIC_ID),
+            role_from_first_message=role_from_first_message,
+            **kwargs,
+        )
+
+
+class PrometheusChannel:
+    """A wrapper for interacting with a prometheus server."""
+
+    def __init__(self, address: Address, metrics: Dict[str, Any], host: str, port: int):
+        """
+        Initialize a prometheus channel.
+
+        :param address: The address of the connection.
+        :param metrics: The prometheus metrics.
+        :param port: The port at which to expose the metrics.
+        """
+        self.address = address
+        self.metrics = metrics
+        self._loop: Optional[asyncio.AbstractEventLoop] = None
+        self._queue: Optional[asyncio.Queue] = None
+        self.logger: Union[logging.Logger, logging.LoggerAdapter] = _default_logger
```

Comment:
logger should be an arg to and only default logger used if no logger passed through